tests/integration/runner.rs

```diff
@@ -0,0 +1,67 @@
+use bootloader::{BootConfig, DiskImageBuilder};
+use std::{
+    io,
+    path::Path,
+    process::{Command, Stdio},
+    thread,
+};
+
+/// Creates a test case that runs a binary form `test_kernel` crate in qemu.
+#[macro_export]
+macro_rules! test {
+    ($bin_name:ident) => {
+        #[test]
+        fn $bin_name() {
+            $crate::runner::run(env!(concat!(
+                "CARGO_BIN_FILE_TEST_KERNEL_",
+                stringify!($bin_name)
+            )));
+        }
+    };
+}
+
+pub fn run(path: &str) {
+    let path = Path::new(path);
+    let mut image_builder = DiskImageBuilder::new(path.to_path_buf());
+    let image_path = path.with_extension(".mbr");
+    image_builder
+        .set_boot_config(&{
+            let mut boot_config = BootConfig::default();
+            boot_config.serial_logging = false;
+            boot_config
+        })
+        .create_bios_image(&image_path)
+        .unwrap();
+
+    let mut child = Command::new("qemu-system-x86_64")
+        .arg("-drive")
+        .arg(format!("format=raw,file={}", image_path.display()))
+        .arg("-device")
+        .arg("isa-debug-exit,iobase=0xf4,iosize=0x04")
+        .arg("-serial")
+        .arg("stdio")
+        .arg("-display")
+        .arg("none")
+        .arg("--no-reboot")
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .stdin(Stdio::null())
+        .spawn()
+        .unwrap();
+
+    let mut child_stdout = child.stdout.take().unwrap();
+    let mut child_stderr = child.stderr.take().unwrap();
+
+    let t1 = thread::spawn(move || io::copy(&mut child_stdout, &mut io::stdout()));
+    let t2 = thread::spawn(move || io::copy(&mut child_stderr, &mut io::stderr()));
+
+    let status = child.wait().unwrap();
+    match status.code() {
+        Some(33) => {}
+        Some(35) => panic!("Test failed"),
```

Comment:
Maybe we should avoid "magic values" and introduce enum with more meaningful names?

Reply:
i tried ...

Reply:
https://os.phil-opp.com/testing/

> The functionality of the isa-debug-exit device is very simple. When a value is written to the I/O port specified by iobase, it causes QEMU to exit with [exit status](https://en.wikipedia.org/wiki/Exit_status) `(value << 1) | 1`. So when we write `0` to the port, QEMU will exit with exit status `(0 << 1) | 1 = 1`, and when we write `1` to the port, it will exit with exit status `(1 << 1) | 1 = 3`.

Reply:
i made an issue for this, but let's get it merged so that i don't block you https://github.com/grupacosmo/cosmos/issues/28